segfault in cm/uct_flush_test.am_pending_flush_nb

http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3232/console
```
16:43:40 [ RUN      ] cm/uct_flush_test.am_pending_flush_nb/0
16:43:42 /var/lib/jenkins/jobs/gh-ucx-pr/workspace/contrib/../test/gtest/uct/test_flush.cc:385: Failure
16:43:42 Value of: flush_req.comp.count == 1
16:43:42   Actual: false
16:43:42 Expected: true
16:43:42 /var/lib/jenkins/jobs/gh-ucx-pr/workspace/contrib/../test/gtest/uct/test_flush.cc:388: Failure
16:43:42 Value of: reqs.back().comp.count
16:43:42   Actual: 2
16:43:42 Expected: 1
16:43:42 /var/lib/jenkins/jobs/gh-ucx-pr/workspace/contrib/../test/gtest/uct/test_flush.cc:388: Failure
16:43:42 Value of: reqs.back().comp.count
16:43:42   Actual: 2
16:43:42 Expected: 1
16:43:42 /var/lib/jenkins/jobs/gh-ucx-pr/workspace/contrib/../test/gtest/uct/test_flush.cc:388: Failure
16:43:42 Value of: reqs.back().comp.count
16:43:42   Actual: 2
16:43:42 Expected: 1
16:43:42 /var/lib/jenkins/jobs/gh-ucx-pr/workspace/contrib/../test/gtest/uct/test_flush.cc:388: Failure
16:43:42 Value of: reqs.back().comp.count
16:43:42   Actual: 2
16:43:42 Expected: 1
16:43:42 /var/lib/jenkins/jobs/gh-ucx-pr/workspace/contrib/../test/gtest/uct/test_flush.cc:388: Failure
16:43:42 Value of: reqs.back().comp.count
16:43:42   Actual: 2
16:43:42 Expected: 1
16:43:42 /var/lib/jenkins/jobs/gh-ucx-pr/workspace/contrib/../test/gtest/uct/test_flush.cc:388: Failure
16:43:42 Value of: reqs.back().comp.count
16:43:42   Actual: 2
16:43:42 Expected: 1
16:43:42 /var/lib/jenkins/jobs/gh-ucx-pr/workspace/contrib/../test/gtest/uct/test_flush.cc:388: Failure
16:43:42 Value of: reqs.back().comp.count
16:43:42   Actual: 2
16:43:42 Expected: 1
16:43:42 /var/lib/jenkins/jobs/gh-ucx-pr/workspace/contrib/../test/gtest/uct/test_flush.cc:388: Failure
16:43:42 Value of: reqs.back().comp.count
16:43:42   Actual: 2
16:43:42 Expected: 1
16:43:42 /var/lib/jenkins/jobs/gh-ucx-pr/workspace/contrib/../test/gtest/uct/test_flush.cc:388: Failure
16:43:42 Value of: reqs.back().comp.count
16:43:42   Actual: 2
16:43:42 Expected: 1
16:43:42 /var/lib/jenkins/jobs/gh-ucx-pr/workspace/contrib/../test/gtest/uct/test_flush.cc:388: Failure
16:43:42 Value of: reqs.back().comp.count
16:43:42   Actual: 2
16:43:42 Expected: 1
16:43:42 [jenkins03:11457:0] Caught signal 11 (Segmentation fault: address not mapped to object at address (nil))
16:43:43 ==== backtrace ====
16:43:43  0 0x000000000000f100 _L_unlock_13()  funlockfile.c:0
16:43:43  1 0x000000000049f9af uct_flush_test::am_send_pending()  /var/lib/jenkins/jobs/gh-ucx-pr/workspace/contrib/../test/gtest/uct/test_flush.cc:102
16:43:43  2 0x000000000049f9af uct_flush_test::am_send_pending()  /var/lib/jenkins/jobs/gh-ucx-pr/workspace/contrib/../test/gtest/uct/test_flush.cc:102
16:43:43  3 0x000000000049f9af uct_flush_test::am_progress()  /var/lib/jenkins/jobs/gh-ucx-pr/workspace/contrib/../test/gtest/uct/test_flush.cc:114
16:43:43  4 0x0000000000065e25 uct_cm_iface_progress()  /var/lib/jenkins/jobs/gh-ucx-pr/workspace/contrib/../src/uct/ib/cm/cm_iface.c:58
16:43:43  5 0x00000000000195fe ucs_callbackq_dispatch()  /var/lib/jenkins/jobs/gh-ucx-pr/workspace/contrib/../src/ucs/datastruct/callbackq.inl:39
16:43:43  6 0x00000000000195fe uct_worker_progress()  /var/lib/jenkins/jobs/gh-ucx-pr/workspace/contrib/../src/uct/base/uct_md.c:231
16:43:43  7 0x0000000000505e53 uct_test::entity::progress()  /var/lib/jenkins/jobs/gh-ucx-pr/workspace/contrib/../test/gtest/uct/uct_test.cc:293
16:43:43  8 0x0000000000505e53 ucs_async_check_miss()  /var/lib/jenkins/jobs/gh-ucx-pr/workspace/contrib/../src/ucs/async/async.h:135
16:43:43  9 0x0000000000505e53 uct_test::entity::async_wrapper::check_miss()  /var/lib/jenkins/jobs/gh-ucx-pr/workspace/contrib/../test/gtest/uct/uct_test.cc:610
16:43:43 10 0x0000000000505e53 uct_test::entity::progress()  /var/lib/jenkins/jobs/gh-ucx-pr/workspace/contrib/../test/gtest/uct/uct_test.cc:294
16:43:43 11 0x0000000000505e53 uct_test::progress()  /var/lib/jenkins/jobs/gh-ucx-pr/workspace/contrib/../test/gtest/uct/uct_test.cc:165
16:43:43 12 0x000000000049d308 uct_flush_test::wait_am()  /var/lib/jenkins/jobs/gh-ucx-pr/workspace/contrib/../test/gtest/uct/test_flush.cc:149
16:43:43 13 0x000000000049d308 uct_flush_test_am_pending_flush_nb_Test::test_body()  /var/lib/jenkins/jobs/gh-ucx-pr/workspace/contrib/../test/gtest/uct/test_flush.cc:394
16:43:43 14 0x0000000000463596 ucs::test_base::run()  /var/lib/jenkins/jobs/gh-ucx-pr/workspace/contrib/../test/gtest/common/test.cc:140
16:43:43 15 0x0000000000463596 ucs::test_base::TestBodyProxy()  /var/lib/jenkins/jobs/gh-ucx-pr/workspace/contrib/../test/gtest/common/test.cc:166
16:43:43 16 0x000000000045b653 HandleSehExceptionsInMethodIfSupported<testing::Test, void>()  /var/lib/jenkins/jobs/gh-ucx-pr/workspace/contrib/../test/gtest/common/gtest-all.cc:3562
16:43:43 17 0x000000000044f67d testing::Test::Run()  /var/lib/jenkins/jobs/gh-ucx-pr/workspace/contrib/../test/gtest/common/gtest-all.cc:3635
16:43:43 18 0x000000000044f74c testing::TestInfo::Run()  /var/lib/jenkins/jobs/gh-ucx-pr/workspace/contrib/../test/gtest/common/gtest-all.cc:3812
16:43:43 19 0x000000000044f8af testing::TestCase::Run()  /var/lib/jenkins/jobs/gh-ucx-pr/workspace/contrib/../test/gtest/common/gtest-all.cc:3930
16:43:43 20 0x00000000004543c7 testing::internal::UnitTestImpl::RunAllTests()  /var/lib/jenkins/jobs/gh-ucx-pr/workspace/contrib/../test/gtest/common/gtest-all.cc:5802
16:43:43 21 0x00000000004546cb testing::internal::UnitTestImpl::RunAllTests()  /var/lib/jenkins/jobs/gh-ucx-pr/workspace/contrib/../test/gtest/common/gtest-all.cc:5719
16:43:43 22 0x000000000040e5e2 RUN_ALL_TESTS()  /var/lib/jenkins/jobs/gh-ucx-pr/workspace/contrib/../test/gtest/common/gtest.h:20058
16:43:43 23 0x0000000000021b15 __libc_start_main()  ???:0
16:43:43 24 0x00000000004396f9 _start()  ???:0
16:43:43 ===================
16:43:43 make: *** [test] Segmentation fault
16:43:43 make: Leaving directory `/scrap/jenkins/jobs/gh-ucx-pr/workspace/build-test/test/gtest'
```

also in http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3289/console
```
17:41:04 [ RUN      ] cm/uct_flush_test.am_pending_flush_nb/1
17:41:05 /var/lib/jenkins/jobs/gh-ucx-pr/workspace-3/contrib/../test/gtest/uct/test_flush.cc:385: Failure
17:41:05 Value of: flush_req.comp.count == 1
17:41:05   Actual: false
17:41:05 Expected: true
17:41:05 /var/lib/jenkins/jobs/gh-ucx-pr/workspace-3/contrib/../test/gtest/uct/test_flush.cc:388: Failure
17:41:05 Value of: reqs.back().comp.count
17:41:05   Actual: 2
17:41:05 Expected: 1
17:41:05 /var/lib/jenkins/jobs/gh-ucx-pr/workspace-3/contrib/../test/gtest/uct/test_flush.cc:388: Failure
17:41:05 Value of: reqs.back().comp.count
17:41:05   Actual: 2
17:41:05 Expected: 1
17:41:05 /var/lib/jenkins/jobs/gh-ucx-pr/workspace-3/contrib/../test/gtest/uct/test_flush.cc:388: Failure
17:41:05 Value of: reqs.back().comp.count
17:41:05   Actual: 2
17:41:05 Expected: 1
17:41:05 /var/lib/jenkins/jobs/gh-ucx-pr/workspace-3/contrib/../test/gtest/uct/test_flush.cc:388: Failure
17:41:05 Value of: reqs.back().comp.count
17:41:05   Actual: 2
17:41:05 Expected: 1
17:41:05 /var/lib/jenkins/jobs/gh-ucx-pr/workspace-3/contrib/../test/gtest/uct/test_flush.cc:388: Failure
17:41:05 Value of: reqs.back().comp.count
17:41:05   Actual: 2
17:41:05 Expected: 1
17:41:05 /var/lib/jenkins/jobs/gh-ucx-pr/workspace-3/contrib/../test/gtest/uct/test_flush.cc:388: Failure
17:41:05 Value of: reqs.back().comp.count
17:41:05   Actual: 2
17:41:05 Expected: 1
17:41:05 /var/lib/jenkins/jobs/gh-ucx-pr/workspace-3/contrib/../test/gtest/uct/test_flush.cc:388: Failure
17:41:05 Value of: reqs.back().comp.count
17:41:05   Actual: 2
17:41:05 Expected: 1
17:41:05 /var/lib/jenkins/jobs/gh-ucx-pr/workspace-3/contrib/../test/gtest/uct/test_flush.cc:388: Failure
17:41:05 Value of: reqs.back().comp.count
17:41:05   Actual: 2
17:41:05 Expected: 1
17:41:05 /var/lib/jenkins/jobs/gh-ucx-pr/workspace-3/contrib/../test/gtest/uct/test_flush.cc:388: Failure
17:41:05 Value of: reqs.back().comp.count
17:41:05   Actual: 2
17:41:05 Expected: 1
17:41:05 /var/lib/jenkins/jobs/gh-ucx-pr/workspace-3/contrib/../test/gtest/uct/test_flush.cc:388: Failure
17:41:05 Value of: reqs.back().comp.count
17:41:05   Actual: 2
17:41:05 Expected: 1
17:41:05 [jenkins03:24101:3] Caught signal 11 (Segmentation fault: address not mapped to object at address (nil))
17:41:06 ==== backtrace ====
17:41:06  0 0x000000000000f100 _L_unlock_13()  funlockfile.c:0
17:41:06  1 0x000000000049ffff uct_flush_test::am_send_pending()  /var/lib/jenkins/jobs/gh-ucx-pr/workspace-3/contrib/../test/gtest/uct/test_flush.cc:102
17:41:06  2 0x000000000049ffff uct_flush_test::am_send_pending()  /var/lib/jenkins/jobs/gh-ucx-pr/workspace-3/contrib/../test/gtest/uct/test_flush.cc:102
17:41:06  3 0x000000000049ffff uct_flush_test::am_progress()  /var/lib/jenkins/jobs/gh-ucx-pr/workspace-3/contrib/../test/gtest/uct/test_flush.cc:114
17:41:06  4 0x0000000000066af5 uct_cm_iface_progress()  /var/lib/jenkins/jobs/gh-ucx-pr/workspace-3/contrib/../src/uct/ib/cm/cm_iface.c:58
17:41:06  5 0x000000000001972e ucs_callbackq_dispatch()  /var/lib/jenkins/jobs/gh-ucx-pr/workspace-3/contrib/../src/ucs/datastruct/callbackq.inl:39
17:41:06  6 0x000000000001972e uct_worker_progress()  /var/lib/jenkins/jobs/gh-ucx-pr/workspace-3/contrib/../src/uct/base/uct_md.c:231
17:41:06  7 0x0000000000506273 uct_test::entity::progress()  /var/lib/jenkins/jobs/gh-ucx-pr/workspace-3/contrib/../test/gtest/uct/uct_test.cc:293
17:41:06  8 0x0000000000506273 ucs_async_check_miss()  /var/lib/jenkins/jobs/gh-ucx-pr/workspace-3/contrib/../src/ucs/async/async.h:135
17:41:06  9 0x0000000000506273 uct_test::entity::async_wrapper::check_miss()  /var/lib/jenkins/jobs/gh-ucx-pr/workspace-3/contrib/../test/gtest/uct/uct_test.cc:610
17:41:06 10 0x0000000000506273 uct_test::entity::progress()  /var/lib/jenkins/jobs/gh-ucx-pr/workspace-3/contrib/../test/gtest/uct/uct_test.cc:294
17:41:06 11 0x0000000000506273 uct_test::progress()  /var/lib/jenkins/jobs/gh-ucx-pr/workspace-3/contrib/../test/gtest/uct/uct_test.cc:165
17:41:06 12 0x000000000049d958 uct_flush_test::wait_am()  /var/lib/jenkins/jobs/gh-ucx-pr/workspace-3/contrib/../test/gtest/uct/test_flush.cc:149
17:41:06 13 0x000000000049d958 uct_flush_test_am_pending_flush_nb_Test::test_body()  /var/lib/jenkins/jobs/gh-ucx-pr/workspace-3/contrib/../test/gtest/uct/test_flush.cc:394
17:41:06 14 0x0000000000463be6 ucs::test_base::run()  /var/lib/jenkins/jobs/gh-ucx-pr/workspace-3/contrib/../test/gtest/common/test.cc:140
17:41:06 15 0x0000000000463be6 ucs::test_base::TestBodyProxy()  /var/lib/jenkins/jobs/gh-ucx-pr/workspace-3/contrib/../test/gtest/common/test.cc:166
17:41:06 16 0x000000000045bde3 HandleSehExceptionsInMethodIfSupported<testing::Test, void>()  /var/lib/jenkins/jobs/gh-ucx-pr/workspace-3/contrib/../test/gtest/common/gtest-all.cc:3562
17:41:06 17 0x000000000044fe0d testing::Test::Run()  /var/lib/jenkins/jobs/gh-ucx-pr/workspace-3/contrib/../test/gtest/common/gtest-all.cc:3635
17:41:06 18 0x000000000044fedc testing::TestInfo::Run()  /var/lib/jenkins/jobs/gh-ucx-pr/workspace-3/contrib/../test/gtest/common/gtest-all.cc:3812
17:41:06 19 0x000000000045003f testing::TestCase::Run()  /var/lib/jenkins/jobs/gh-ucx-pr/workspace-3/contrib/../test/gtest/common/gtest-all.cc:3930
17:41:06 20 0x0000000000454b57 testing::internal::UnitTestImpl::RunAllTests()  /var/lib/jenkins/jobs/gh-ucx-pr/workspace-3/contrib/../test/gtest/common/gtest-all.cc:5802
17:41:06 21 0x0000000000454e5b testing::internal::UnitTestImpl::RunAllTests()  /var/lib/jenkins/jobs/gh-ucx-pr/workspace-3/contrib/../test/gtest/common/gtest-all.cc:5719
17:41:06 22 0x000000000040e582 RUN_ALL_TESTS()  /var/lib/jenkins/jobs/gh-ucx-pr/workspace-3/contrib/../test/gtest/common/gtest.h:20058
17:41:06 23 0x0000000000021b15 __libc_start_main()  ???:0
17:41:06 24 0x0000000000439e8d _start()  ???:0
17:41:06 ===================
17:41:06 make: *** [test] Segmentation fault
``

--- a/test/gtest/uct/test_dc.cc
+++ b/test/gtest/uct/test_dc.cc
@@ -468,8 +468,9 @@ UCS_TEST_P(test_dc_flow_control, dci_leak)
     EXPECT_UCS_OK(uct_ep_pending_add(m_e1->ep(0), &req));
 
     /* Make sure that ep does not hold dci when sends completed */
-    ucs_time_t timeout    = ucs_get_time() + ucs_time_from_sec(10);
     uct_dc_iface_t *iface = ucs_derived_of(m_e1->iface(), uct_dc_iface_t);
+    ucs_time_t timeout    = ucs_get_time() +
+                            ucs_time_from_sec(UCT_TEST_TIMEOUT_IN_SEC);
     while (iface->tx.stack_top && (ucs_get_time() < timeout)) {
         progress();
     }

--- a/test/gtest/uct/test_flush.cc
+++ b/test/gtest/uct/test_flush.cc
@@ -373,7 +373,8 @@ UCS_TEST_P(uct_flush_test, am_pending_flush_nb) {
      }
 
      /* timeout used to prevent test hung */
-     ucs_time_t loop_end_limit_comp = ucs_get_time() + ucs_time_from_sec(1.0);
+     ucs_time_t loop_end_limit_comp = ucs_get_time() +
+                                      ucs_time_from_sec(UCT_TEST_TIMEOUT_IN_SEC);
      /* coverity[loop_condition] */
      while (flush_req.comp.count != 1) {
          if (ucs_get_time() > loop_end_limit_comp) {
@@ -389,9 +390,9 @@ UCS_TEST_P(uct_flush_test, am_pending_flush_nb) {
          reqs.pop_back();
      }
 
-     sender().destroy_ep(0);
-
      wait_am(count);
+
+     sender().destroy_ep(0);
 
      uct_iface_set_am_handler(receiver().iface(), AM_ID, NULL, NULL, 0);
 

--- a/test/gtest/uct/test_rc.cc
+++ b/test/gtest/uct/test_rc.cc
@@ -83,8 +83,9 @@ ucs_status_t test_rc_flow_control::am_send(uct_pending_req_t *self)
 
 void test_rc_flow_control::validate_grant(entity *e)
 {
-    ucs_time_t timeout = ucs_get_time() + ucs_time_from_sec(10.0);
-    while ((ucs_get_time() < timeout) && (get_fc_ptr(e)->fc_wnd <= 0)) {
+    ucs_time_t timeout = ucs_get_time() +
+                         ucs_time_from_sec(UCT_TEST_TIMEOUT_IN_SEC);
+    while ((ucs_get_time() < timeout) && (!get_fc_ptr(e)->fc_wnd)) {
         short_progress_loop();
     }
     EXPECT_GT(get_fc_ptr(e)->fc_wnd, 0);

--- a/test/gtest/uct/test_stats.cc
+++ b/test/gtest/uct/test_stats.cc
@@ -64,8 +64,12 @@ public:
     void check_am_rx_counters(size_t len) {
         uint64_t v;
 
-        short_progress_loop(100.0);
-        v = UCS_STATS_GET_COUNTER(uct_iface(receiver())->stats, UCT_IFACE_STAT_RX_AM);
+        ucs_time_t timeout = ucs_get_time() + ucs_time_from_sec(UCT_TEST_TIMEOUT_IN_SEC);
+        do {
+            short_progress_loop();
+            v = UCS_STATS_GET_COUNTER(uct_iface(receiver())->stats, UCT_IFACE_STAT_RX_AM);
+        } while ((ucs_get_time() < timeout) && !v);
+
         EXPECT_EQ(1UL, v);
         v = UCS_STATS_GET_COUNTER(uct_iface(receiver())->stats, UCT_IFACE_STAT_RX_AM_BYTES);
         EXPECT_EQ(len, v);
@@ -216,12 +220,18 @@ UCS_TEST_P(test_uct_stats, get_bcopy)
 
     check_caps(UCT_IFACE_FLAG_GET_BCOPY);
 
+    m_comp.count = 2;
+    m_comp.func  = NULL;
     status = uct_ep_get_bcopy(sender_ep(), (uct_unpack_callback_t)memcpy,
                               lbuf->ptr(), lbuf->length(),
-                              rbuf->addr(), rbuf->rkey(), NULL);
+                              rbuf->addr(), rbuf->rkey(), &m_comp);
     EXPECT_TRUE(UCS_INPROGRESS == status || UCS_OK == status);
 
-    short_progress_loop();
+    ucs_time_t timeout = ucs_get_time() + ucs_time_from_sec(UCT_TEST_TIMEOUT_IN_SEC);
+    do {
+        short_progress_loop();
+    } while ((ucs_get_time() < timeout) && (m_comp.count != 1));
+
     check_tx_counters(UCT_EP_STAT_GET, UCT_EP_STAT_BYTES_BCOPY,
                       lbuf->length());
 }
@@ -235,8 +245,15 @@ UCS_TEST_P(test_uct_stats, get_zcopy)
     UCS_TEST_GET_BUFFER_IOV(iov, iovcnt, lbuf->ptr(), lbuf->length(), lbuf->memh(),
                             sender().iface_attr().cap.get.max_iov);
 
+    m_comp.count = 2;
+    m_comp.func  = NULL;
     status = uct_ep_get_zcopy(sender_ep(), iov, iovcnt, rbuf->addr(), rbuf->rkey(), 0);
     EXPECT_TRUE(UCS_INPROGRESS == status || UCS_OK == status);
+
+    ucs_time_t timeout = ucs_get_time() + ucs_time_from_sec(UCT_TEST_TIMEOUT_IN_SEC);
+    do {
+        short_progress_loop();
+    } while ((ucs_get_time() < timeout) && (m_comp.count != 1));
 
     short_progress_loop();
     check_tx_counters(UCT_EP_STAT_GET, UCT_EP_STAT_BYTES_ZCOPY,

--- a/test/gtest/uct/uct_test.h
+++ b/test/gtest/uct/uct_test.h
@@ -17,6 +17,9 @@ extern "C" {
 #include <vector>
 
 
+#define UCT_TEST_TIMEOUT_IN_SEC   10.0
+
+
 /* Testing resource */
 struct resource {
     virtual ~resource() {};

